### PR TITLE
Add fastem support for non-rectangular in the acquisition manager.

### PR DESCRIPTION
**Pull request by BasHolweg**
Add the functionality to be able to acquire non-rectangular or rotated rectangular regions of acquisition (ROA's). For now the previous method is still used for acquisition since the GUI still requires compatibility. Two utility functions were added for flood filling a ndarray, which is used for finding the inside of the polygonal ROA's and for finding the bounding box of a set of polygon coordinates.

### CODE REVIEW

File: src/odemis/acq/fastem.py [hunk 1]
```diff
@@ -217,6 +222,143 @@ def _calculate_field_indices(self):
 
         return field_indices
 
+    # TODO: should not require input argument but use self.coordinates when the behaviour is adjusted to support
+    #  multiple coordinate points in stead of 4 separate values
```
Comment (nandishjpatel) [inline src/odemis/acq/fastem.py:226]:
```suggestion
    #  multiple coordinate points instead of 4 separate values
```

File: src/odemis/acq/fastem.py [hunk 2]
```diff
@@ -155,6 +154,12 @@ def estimate_acquisition_time(self):
         return tot_time
 
     def _calculate_field_indices(self):
+        # TODO: Old method is used since polygonal fields is not yet supported by the gui. This must be adjusted when
+        #  it is.
+        indices = self.get_square_field_indices(self.coordinates.value)
+        return indices
+
+    def get_square_field_indices(self, coordinates):
```
Comment (nandishjpatel) [inline src/odemis/acq/fastem.py:162]: `:param coordinates:` left to be stated in docstring

File: src/odemis/acq/fastem.py [hunk 3]
```diff
@@ -217,6 +222,143 @@ def _calculate_field_indices(self):
...
+        :return: (ndarray) Filled numpy array that represent the fields that need to be acquired.
+        """
+        # Get the bounding megafield shape by taking the maximum indices from the indices list and create a boolian
+        # numpy array to represent the megafield with True values for fields that need to be acquired
+        _, _, rmax, cmax = util.get_polygon_bbox(indices)
+        megafield_grid_shape = (rmax + 1, cmax + 1)
+        megafield_grid_rep = numpy.zeros(megafield_grid_shape, dtype=numpy.bool)
+        megafield_grid_rep[[r[0] for r in indices], [c[1] for c in indices]] = True
+
+        # Pad the array with zeros so the flood fill wil always surround the entire shape
```
Comment (nandishjpatel) [inline src/odemis/acq/fastem.py:347]:
```suggestion
        # Pad the array with zeros so the flood fill will always surround the entire shape
```

File: src/odemis/util/img.py [hunk 4]
```diff
@@ -1201,3 +1201,41 @@ def assembleZCube(images, zlevels):
...
+    overflow_counter = 0
+    while len(queue) > 0:
+        # To prevent the possibility of an infinite loop count the number of iterations. If it is larger than the
+        # total number of values in the array raise an error since this should not be possible.
+        if overflow_counter > max_area:
+            raise ValueError(f"Number of loop iterations is higher than maximum iterations possible ({max_area}) in "
+                             f"array of shape: {array.shape}")
+
+        row, col = queue.pop(0)  # Set the current pixel to the first element of the queue and remove it from the queue
+        if not array[row, col]:
```
Comment (nandishjpatel) [inline src/odemis/util/img.py:1230]: If I understood correctly, I think the below code can be optimized reference: https://www.baeldung.com/cs/flood-fill-algorithm#2-naive-implementation

File: src/odemis/acq/fastem.py [hunk 5]
```diff
@@ -217,6 +222,143 @@ def _calculate_field_indices(self):
...
+
+    @staticmethod
+    def _create_and_fill_megafield_rep(indices):
+        """
+        Create a numpy array that represents the given indexes as True values and fills any enclosed False values.
+
+        :param indices: (list of nested tuples (row,column)) Index values of the shape that needs to be filled.
+        :return: (ndarray) Filled numpy array that represent the fields that need to be acquired.
+        """
+        # Get the bounding megafield shape by taking the maximum indices from the indices list and create a boolian
```
Comment (tepals) [inline src/odemis/acq/fastem.py:340]:
```suggestion
        # Get the bounding megafield shape by taking the maximum indices from the indices list and create a boolean 
```

File: src/odemis/util/img.py [hunk 6]
```diff
@@ -1201,3 +1201,41 @@ def assembleZCube(images, zlevels):
         ret = DataArray(ret, metadata3d)
 
         return ret
+
+
+def apply_flood_fill(array, start):
+    """
+    Flood fills a 2-Dimensional numpy array with truth values from a given start position with the 4-connected method.
+    The input array contains a shape represented with truth values for the borders of the shape.
+
```
Comment (tepals) [inline src/odemis/util/img.py:1210]: could you give a small example drawing? Something like

```python
# input array                output array
# [[0, 0, 0, 0, 0, 0,],     [[0, 0, 0, 0, 0, 0,],
#  [0, 0, 1, 1, 0, 0,],      [0, 0, 1, 1, 0, 0,],
#  [0, 1, 0, 0, 1, 0,],  ->  [0, 1, 1, 1, 1, 0,],
#  [0, 0, 1, 1, 1, 0,],      [0, 0, 1, 1, 1, 0,],
#  [0, 0, 0, 0, 0, 0,]]      [0, 0, 0, 0, 0, 0,]]
```

File: src/odemis/util/img.py [hunk 7]
```diff
@@ -1201,3 +1201,41 @@ def assembleZCube(images, zlevels):
...
+    :param array: (ndarray(bool)) binary array of size mxn containing the shape to be filled
+    :param start: (tuple(int, int)) position from which to start flood fill
+    :return: (ndarray(bool)) array of size mxn containing the filled shape, with filled values True
+    """
+    if start >= array.shape:
+        raise ValueError(f"Start position of {start} does not lie within the array.")
+
+    array = array.copy()
```
Comment (tepals) [inline src/odemis/util/img.py:1218]: `array` is a python standard module, I think it's good to be a bit careful with naming things the same as other libraries. I would suggest to rename this, maybe to something like `input_array` or `shape_array`

File: src/odemis/util/img.py [hunk 8]
```diff
@@ -1201,3 +1201,41 @@ def assembleZCube(images, zlevels):
...
+    max_area = array.shape[0] * array.shape[1]
+    queue = [start]
```
Comment (tepals) [inline src/odemis/util/img.py:1220]: `queue` is a library that's used in other places in Odemis, I would suggest to rename this, maybe to `check_pixel`

File: src/odemis/util/test/img_test.py [hunk 9]
```diff
@@ -1181,6 +1181,40 @@ def test_rotate_img_metadata(self):
             numpy.testing.assert_array_equal(image, rotated_img)  # The data should not change, only the metadata.
 
 
+class TestFloodFill(unittest.TestCase):
+
+    def test_standard_fill(self):
+        """Test the flood_fill as it is expected to be used"""
```
Comment (tepals) [inline src/odemis/util/test/img_test.py:1187]: I would say"

```suggestion
        """Test the expected use of flood_fill"""
```

File: src/odemis/util/test/img_test.py [hunk 10]
```diff
@@ -1181,6 +1181,40 @@ def test_rotate_img_metadata(self):
...
+        for i in range(min(a.shape)):
+            a[i, i] = True
+            a[-i - 1, i] = True
+
+        filled_array = img.apply_flood_fill(a, (3, 0))
+
+        self.assertTrue(filled_array[3, 1])
+        self.assertFalse(filled_array[3, -1])
+        self.assertFalse(filled_array[0, 3])
+        self.assertFalse(filled_array[-1, 3])
```
Comment (tepals) [inline src/odemis/util/test/img_test.py:1198]: I think you could do a more thorough check here. You know exactly what the filled array should look like, so you could just create the expected filled array and check that they are equal.

```python
expected_array = numpy.array([
       [1, 0, 0, 0, 0, 0, 1],
       [1, 1, 0, 0, 0, 1, 0],
       [1, 1, 1, 0, 1, 0, 0],
       [1, 1, 1, 1, 0, 0, 0],
       [1, 1, 1, 0, 1, 0, 0],
       [1, 1, 0, 0, 0, 1, 0],
       [1, 0, 0, 0, 0, 0, 1]
])
numpy.testing.assert_array_equal(expected_array, filled_array)
```

File: src/odemis/util/test/util_test.py [hunk 11]
```diff
@@ -543,6 +543,20 @@ def test_expand_rect(self):
                                                        "expanded rectangle %s" % (expanded_rect, expected_rect))
 
 
+class GetMinMaxTupleTestCase(unittest.TestCase):
+
+    def test_minmax(self):
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:548]: can you rename these to match the new function name?

File: src/odemis/util/test/util_test.py [hunk 12]
```diff
@@ -543,6 +543,20 @@ def test_expand_rect(self):
...
+        """Test getting the minimum and maximum values from a list of tuples"""
```
Comment (tepals) [inline src/odemis/util/test/util_test.py:549]: can you add a test case that checks what happens when you give the function negative values?

File: src/odemis/util/__init__.py [hunk 13]
```diff
@@ -389,6 +389,20 @@ def expand_rect(rect, margin):
     return minx - margin, miny - margin, maxx + margin, maxy + margin
 
 
+def get_polygon_bbox(coordinates):
+    """
+    Get the maximum and minimum values for a and b from a list of tuples
+    with shape: [(a1,b1), (a2,b2), ....., (an,bn)]
+
+    :param coordinates: (list of nested tuples (a,b))
```
Comment (tepals) [inline src/odemis/util/__init__.py:397]: In your implementation coordinates can also be a single tuple, for instance `coordinates = [(0, 2)]`. The output is then `[0, 2, 0, 2]`, I think it would make sense to raise an error when the input is a single coordinate.

Additionally the input tuples can also be more than 2 values long, I think it would also make sense to check for that or mention in the docstring that only the first two values of each tuple are used. For instance you can have `coordinates = [(0, 1, 2), (0, 3)]`, the output is then `(0, 1, 0, 3)`.

**Review by tepals — CHANGES_REQUESTED** [review]

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/util/__init__.py (hunk 13 above)
Comment (BasHolweg) [reply src/odemis/util/__init__.py:397]: That functionality has unfortunately been added since python 3.10.

File: src/odemis/acq/test/fastem_test.py [hunk 14]
```diff
@@ -200,9 +200,9 @@ def test_estimate_acquisition_time(self):
 
             self.assertAlmostEqual(estimated_roa_acq_time, roa_acq_time)
 
-    def test_calculate_field_indices(self):
+    def test_get_square_field_indices(self):
         """Check that the correct number and order of field indices is returned and that row and column are in the
-        correct order."""
+                correct order."""
```
[the remark this answers is not in the document]
Comment (BasHolweg) [reply src/odemis/acq/test/fastem_test.py:205]: It was not a conscious change, I think pycharm reformatted it for me. Have changed it back now

File: src/odemis/acq/fastem.py (hunk 2 above)
Comment (tepals) [reply src/odemis/acq/fastem.py:162]: don't forget this comment

File: src/odemis/acq/fastem.py [hunk 15]
```diff
@@ -217,6 +222,139 @@ def _calculate_field_indices(self):
...
+
+        :param indices: (list of nested tuples (row,column)) Index values of the shape that needs to be filled.
+        :return: (ndarray) Filled numpy array that represent the fields that need to be acquired.
+        """
+        # Get the bounding megafield shape by taking the maximum indices from the indices list and create a boolean
+        # numpy array to represent the megafield with True values for fields that need to be acquired
+        _, _, rmax, cmax = util.get_polygon_bbox(indices)
+        megafield_grid_shape = (rmax + 1, cmax + 1)
+        megafield_grid_rep = numpy.zeros(megafield_grid_shape, dtype=numpy.bool)
+        megafield_grid_rep[[r[0] for r in indices], [c[1] for c in indices]] = True
```
Comment (tepals) [inline src/odemis/acq/fastem.py:341]: this would also work:
```
megafield_grid_rep[numpy.array(indices)[:, 0], numpy.array(indices)[:, 1] = True
```
pick whichever you prefer

**Review by tepals — COMMENTED** [review]
looks good!

**Review by tepals — APPROVED** [review]

**Review by nandishjpatel — APPROVED** [review]
